src/selective-reader.js

```diff
@@ -0,0 +1,101 @@
+//
+// Copyright 2019 DXOS.org
+//
+
+import { Readable } from 'stream';
+import createBatchStream from './create-batch-stream';
+
+/**
+ * Creates a multi ReadableStream for feed streams.
+ */
+export default class SelectiveReader extends Readable {
+  /** @type {(feedDescriptor, message) => Promise<boolean>} */
+  _evaluator;
+
+  /** @type {Set<{ descriptor: FeedDescriptor, stream: any, buffer: any[] }>} */
+  _feeds = new Set();
+
+  /** @type {() => void} */
+  _wakeUpReader;
+
+  /** @type {Promise} */
+  _hasData;
+
+  _reading = false;
+
+  constructor (evaluator) {
+    super({ objectMode: true });
+
+    this._evaluator = evaluator;
+    this._resetDataLock();
+  }
+
+  _resetDataLock () {
+    this._hasData = new Promise(resolve => { this._wakeUpReader = resolve; });
+  }
+
+  async _read () {
+    if (this._reading) {
+      this._needsData = true;
+      return;
+    }
+    this._reading = true;
+    this._needsData = false;
+
+    while (true) {
+      this._resetDataLock();
+
+      for (const feed of this._feeds.values()) {
+        if (feed.buffer.length === 0) {
+          const messages = feed.stream.read();
+          if (!messages) continue;
+          feed.buffer.push(...messages);
```

Comment:
Do we know what if anything limits how large `buffer` can grow?
Consider adding a test for this (write large amounts of data without reading).

Reply:
Buffer should at max be the size of a chunk that `stream.read` returns since the next read occurs only only when the buffer was emptied. In worst case that's the stream's entire in-memory buffer.

Reply:
The actual size should depend on feed reader's chunk size